### PR TITLE
feat(tExtRef): Export function to lookup control block from ExtRef

### PR DESCRIPTION
**Pull request by danyill**
I rewrote the function, added tests and exported it.

I will be most pleased to take a review.

### CODE REVIEW

**Comment by danyill** [comment]
So perhaps the tests are not quite as comprehensive as they could be, but with the old function I have some test failures (11!).

IMHO this is a better function because it just uses the querySelector without the find which should be more performant.

I am unsure about:

* whether `serviceType` should be required which could simplify this a little more but for now I have allowed it to be missing.
* the way I have handled missing attributes vs. empty attributes (I find the standard very unclear in this regard).

I will be most pleased to make improvements.

**Review by JakobVogelsang — CHANGES_REQUESTED** [review]
I have not looked through the test file and the tests yet. I do agree with you that your approach is potentially more performant. We need to conserved more edge cases though, such as `srcPrefix` is empty string and the `prefix` of the logical node is not existing. Same I think is true for inst.

File: tExtRef/sourceControlBlock.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  const [iedName, srcPrefix, srcLNInst, srcCBName] = [
+    "iedName",
+    "srcPrefix",
+    "srcLNInst",
+    "srcCBName",
+  ].map((attr) => extRef.getAttribute(attr));
+
+  const doc = extRef.ownerDocument;
+  const srcLDInst =
+    extRef.getAttribute("srcLDInst") ?? extRef.getAttribute("ldInst");
```
[the remark this answers is not in the document]
Comment (danyill) [reply tExtRef/sourceControlBlock.ts:19]: Table 33 of Part 6 Ed 2.1 says for `srcLDInst`, 

> The LD inst of the source control block – if missing, same as ldInst above. 

Given that `getAttribute` will return `null` if `ldInst` does not exist, I am not sure how this can return anything other than `null` if both attributes are missing. Can you explain?

File: tExtRef/sourceControlBlock.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  const srcLNClass = extRef.getAttribute("srcLNClass") ?? "LLN0";
```
[the remark this answers is not in the document]
Comment (danyill) [reply tExtRef/sourceControlBlock.ts:20]: Again, Table 33 of the standard says:

> The LN class of the LN, where the source control block resides; if missing, LLN0

So I think it is correct but happy to take a contrary opinion :wink:

File: tExtRef/sourceControlBlock.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  if (!iedName || !srcLDInst || !srcCBName || serviceType === "Poll")
+    return null;
+
+  const lDevice = `:root > IED[name="${iedName}"] > AccessPoint > Server > LDevice[inst="${srcLDInst}"]`;
+
+  const maybeReport = !serviceType || serviceType === "Report";
+  const maybeGSE = !serviceType || serviceType === "GOOSE";
+  const maybeSV = !serviceType || serviceType === "SMV";
+
+  const lN = maybeReport && srcLNClass !== "LLN0" ? "LN" : "LN0";
```
[the remark this answers is not in the document]
Comment (danyill) [reply tExtRef/sourceControlBlock.ts:33]: I guess it's not. Given that `SampledValueControl` and `GSEControl` must be in `LLN0`, my thinking was that only if it's a report can it be a `LN`. However I think you are correct and it is not strictly necessary because `srcLNClass` is always defined or the default.

File: tExtRef/sourceControlBlock.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    extRef.getAttribute("serviceType") ?? extRef.getAttribute("pServT");
+
+  if (!iedName || !srcLDInst || !srcCBName || serviceType === "Poll")
+    return null;
+
+  const lDevice = `:root > IED[name="${iedName}"] > AccessPoint > Server > LDevice[inst="${srcLDInst}"]`;
+
+  const maybeReport = !serviceType || serviceType === "Report";
+  const maybeGSE = !serviceType || serviceType === "GOOSE";
+  const maybeSV = !serviceType || serviceType === "SMV";
```
[the remark this answers is not in the document]
Comment (danyill) [reply tExtRef/sourceControlBlock.ts:31]: Sure, even `maybeSMV` !

File: tExtRef/sourceControlBlock.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+  const lN = maybeReport && srcLNClass !== "LLN0" ? "LN" : "LN0";
+
+  const lNQualifiers = [
```
[the remark this answers is not in the document]
Comment (danyill) [reply tExtRef/sourceControlBlock.ts:35]: Sounds fair, thank you for this direction.

Let's see how we go. It might be fun to try and benchmark this because performance-wise this causes quite a rapid deterioration if the Extref is not "fully qualified".

Having said that, the selection is fully qualified back to the root node so it should still be quite performant. Really must write some benchmarks to see if any of this speculation is true :wink:

**Comment by danyill** [comment]
Thank you for this most helpful review.

I have a couple of questions and I need to write some more tests (I've created stubs but haven't written) them.

If you have a moment to tell me if it looks like it's heading in the right direction I'll fill out the stubs and add any other obviously missing tests.

**Comment by danyill** [comment]
Just because I had some time, I had a quick hack to look at performance:

set `this!.slow(0)`;

Run 10 tests with old code and new code.
Imagine failing tests with old code don't matter.

Record times:

Old: 

* mean = 76 ms
* std deviation = 40 ms

New

* mean = 83 ms
* std deviation = 24 ms

This is probably not worth much. Tests are probably not the best performance measure and this was done in a wholly unscientific fashion :wink: 

So I tested on a large scd file iterating over a pile of ExtRefs with both the old and new code:

With 387 ExtRefs, of which 223 have a subscription.

- New code: 3364 ms
- Old code: 8128 ms

So I think we're on the right track.
I expect the old code could perhaps be improved substantially by just fully qualifying the  locations but actually it made it worse (8454 ms)!

**Comment by danyill** [comment]
> We need to conserved more edge cases though, such as srcPrefix is empty string and the prefix of the logical node is not existing. Same I think is true for inst.

For inst, this is required for LN and is set to a fixed value of the empty string for LN0.  So what the standard means (in at least this case) by "if missing, no prefix", means a prefix of "".  I'm almost tempted to apply this interpretation to `inst` but I think it's safest to do as you suggest.

**Comment by danyill** [comment]
> We need to conserved more edge cases though, such as srcPrefix is empty string and the prefix of the logical node is not existing. Same I think is true for inst.
> 
>> For inst, this is required for LN and is set to a fixed value of the empty string for LN0. So what the standard means (in at least this case) by "if missing, no inst", means a inst of "". I'm almost tempted to apply this interpretation to prefix but I think it's safest to do as you suggest.

Nope, I am quite wrong -- this is not really true. There is something subtle here, more in a bit.

File: tExtRef/sourceControlBlock.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply tExtRef/sourceControlBlock.ts:21]: You are quite right!!! AND I did not know AND is one of the things of IEC 61850 that completely surprises me as the two do not have much in common. When you want to separate data from it carries and in another place you assume otherwise ... not very good modelling. 

BUT: I would stick to the standard as this is the smallest chance for false positives.

Thank you for the hint

File: tExtRef/sourceControlBlock.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (JakobVogelsang) [reply tExtRef/sourceControlBlock.ts:20]: You are quite right!!! AND I did not know AND is one of the things of IEC 61850 that completely surprises me as the two do not have much in common. When you want to separate data from it carries and in another place you assume otherwise ... not very good modelling. 

BUT: I would stick to the standard as this is the smallest chance for false positives.

Thank you for the hint

File: tExtRef/sourceControlBlock.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  const maybeSMV = !serviceType || serviceType === "SMV";
+
+  const lN = srcLNClass === "LLN0" ? "LN0" : "LN";
+
+  const lnClass = `[lnClass="${srcLNClass}"]`;
+
+  const lnInst =
+    srcLNInst && srcLNInst !== "" ? `[inst="${srcLNInst}"]` : '[inst=""]';
+
+  // missing or empty string for attribute implies
```
Comment (JakobVogelsang) [inline tExtRef/sourceControlBlock.ts:41]: This comment is unclear to me. It implies that you want to handle `inst` the same way you want to do with `prefix`. The code does behave differently.

**Comment by JakobVogelsang** [comment]
Sorry, I did not read right away to hold off with the review. I general, I do think that performance for this function matters. I am general, therefore, for this changed approach. It is a pleasure to work with you. Doing performance test just because you care. LOVE IT!

**Comment by danyill** [comment]
OK, clarity emerges:

> We need to conserved more edge cases though, such as srcPrefix is empty string and the prefix of the logical node is not existing. Same I think is true for inst.
> 
>> For inst, this is required for LN and is set to a fixed value of the empty string for LN0. So what the standard means (in at least this case) by "if missing, no inst", means a inst of "". I'm almost tempted to apply this interpretation to prefix but I think it's safest to do as you suggest.
>> 
>>>
>>>Nope, I am quite wrong -- this is not really true. There is something subtle here, more in a bit.
>>>
>>>Perhaps best to hold off review for a bit and I will finish in the next < 24 hrs.

OK, for `inst` I think this is where I'm at:

* For `LN0` `inst` must be set to exactly `""`. It may not be a missing attribute
* For `LN` `inst` must be set to a value and may not be an empty string.

Same applies to `srcLNInst`

With `prefix`, it is the odd one out.

* For `LN0` no prefix is allowed
* For `LN` it is optional and so it could be missing or the empty string so we must check both.

**Comment by danyill** [comment]
My prefix checking logic is giving me quite a speed up so while it looks like it might not be necessary, for LNs (only) and those with without a srcPrefix we must do twice as much work so it's worth checking for them.

```ts
if (anyLN === "LN") {
    lnPrefixQualifiers =
      srcPrefix && srcPrefix !== ""
        ? [`[prefix="${srcPrefix}"]`]
        : [":not([prefix])", '[prefix=""]'];
  } else {
    lnPrefixQualifiers = [":not([prefix])"];
  }
```

**Comment by danyill** [comment]
I have corrected a bit of my thinking (this referencing is somewhat inconsistent in its approach and semantics) and I think I've covered a reasonable set of courses.

I shall be delighted for further review and improvement suggestions.

File: tExtRef/sourceControlBlock.spec.ts [hunk 7]
```diff
@@ -0,0 +1,448 @@
+import { expect } from "chai";
+
+import { sourceControlBlock } from "./sourceControlBlock.js";
+import { sourceControlBlockDoc } from "./sourceControlBlock.testfiles.js";
+import { findElement } from "../foundation/helpers.test.js";
+
+const testDoc = findElement(sourceControlBlockDoc) as XMLDocument;
+
+describe("Function to return a control block from an ExtRef", function () {
+  this!.slow(0);
```
Comment (JakobVogelsang) [inline tExtRef/sourceControlBlock.spec.ts:10]: Is this still necessary

Comment (JakobVogelsang) [reply tExtRef/sourceControlBlock.spec.ts:10]: ?

**Review by JakobVogelsang — DISMISSED** [review]
LGTM. I think the `slow` need to be removed.

File: tExtRef/sourceControlBlock.spec.ts (hunk 7 above)
Comment (danyill) [reply tExtRef/sourceControlBlock.spec.ts]: Quite right. Thanks.

**Comment by danyill** [comment]
Thank you, changes done and version updated in `package.json`.

**Review by JakobVogelsang — APPROVED** [review]
LGTM
